destination/destination_integration_test.go

```diff
@@ -0,0 +1,395 @@
+	is.Equal(ok, true)
+	is.Equal(int(num), testNumberDataChange)
+
+	_, err = col.DeleteMany(ctx, bson.M{})
+	is.NoErr(err)
+}
+
+func TestDestination_Write_updateFailureNoKeys(t *testing.T) {
+	is := is.New(t)
+
+	cfg := prepareConfig(t)
+
+	destination := NewDestination()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := destination.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	err = destination.Open(ctx)
+	is.NoErr(err)
+
+	col, err := getTestCollection(ctx, cfg[config.KeyCollection])
+	is.NoErr(err)
+
+	t.Cleanup(func() {
+		err = col.Drop(context.Background())
+		is.NoErr(err)
+
+		err = destination.Teardown(ctx)
+		is.NoErr(err)
+	})
+
+	n, err := destination.Write(ctx, []sdk.Record{getTestCreateRecord(t)})
+	is.NoErr(err)
+	is.Equal(n, 1)
+
+	_, err = destination.Write(ctx, []sdk.Record{getTestUpdateRecordNoKeys(t)})
+	is.True(errors.Is(err, writer.ErrEmptyKey))
+
+	_, err = col.DeleteMany(ctx, bson.M{})
+	is.NoErr(err)
+}
+
+func TestDestination_Write_deleteSuccess(t *testing.T) {
+	is := is.New(t)
+
+	cfg := prepareConfig(t)
+
+	destination := NewDestination()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	err := destination.Configure(ctx, cfg)
+	is.NoErr(err)
+
+	err = destination.Open(ctx)
+	is.NoErr(err)
+
+	col, err := getTestCollection(ctx, cfg[config.KeyCollection])
+	is.NoErr(err)
+
+	t.Cleanup(func() {
+		err = col.Drop(context.Background())
+		is.NoErr(err)
+
+		err = destination.Teardown(ctx)
+		is.NoErr(err)
+	})
+
+	n, err := destination.Write(ctx, []sdk.Record{getTestCreateRecord(t)})
+	is.NoErr(err)
+	is.Equal(n, 1)
+
+	n, err = destination.Write(ctx, []sdk.Record{getTestDeleteRecord(t)})
+	is.NoErr(err)
+	is.Equal(n, 1)
+
+	res, err := col.Find(ctx, bson.M{})
+	is.NoErr(err)
+	is.True(!res.Next(ctx))
+
+	_, err = col.DeleteMany(ctx, bson.M{})
+	is.NoErr(err)
+}
+
+func getTestCollection(ctx context.Context, collection string) (*mongo.Collection, error) {
+	conn, err := mongo.Connect(ctx, options.Client().ApplyURI(testURI))
+	if err != nil {
+		return nil, fmt.Errorf("connect to mongo: %w", err)
+	}
+
+	return conn.Database(testDB).Collection(collection), nil
+}
+
+func getTestCreateRecord(t *testing.T) sdk.Record {
+	t.Helper()
+
+	return sdk.Util.Source.NewRecordCreate(
+		nil, nil,
+		// in insert keys are not used, so we can omit it
+		nil,
+		sdk.StructuredData{
+			testIDFieldName:     testID, // we put it as string here, codec will translate it into ObjectID
+			testTextFieldName:   testTextData,
+			testNumberFieldName: testNumberData,
+		},
+	)
+}
+
+func getTestSnapshot(t *testing.T) sdk.Record {
+	t.Helper()
+
+	return sdk.Util.Source.NewRecordSnapshot(
+		nil, nil,
+		// in insert keys are not used, so we can omit it
+		nil,
+		sdk.StructuredData{
+			testIDFieldName:     testID, // we put it as string here, codec will translate it into ObjectID
+			testTextFieldName:   testTextData,
+			testNumberFieldName: testNumberData,
+		},
+	)
+}
```

Comment:
I'd rather do the following (something similar in all the functions that create test records):
```go
func getTestSnapshot(t *testing.T) (sdk.Record, sdk.StructuredData) {
	t.Helper()

	payloadAfter := sdk.StructuredData{
		testIDFieldName:   primitive.NewObjectID(),
		testTextFieldName: gofakeit.Name(),
	}

	return sdk.Util.Source.NewRecordSnapshot(nil, nil, nil, payloadAfter), payloadAfter
}
```

And let's not use the integer field, as it is marshaled as a float when we deal with `map[string]any` and we need to convert it to integer again.

Reply:
changed logic, maybe we will not gonna need it